Alias direct_descendants as subclasses

jonathanhefner opened the pull request:
This enables e.g. `klass.subclasses` and `klass.direct_descendants` both to work regardless of whether `klass` extends `DescendantsTracker`, and to get a performance boost when `klass` *does* extend `DescendantsTracker`.

jonathanhefner committed:
Alias direct_descendants as subclasses

This alias overrides Class#subclasses.  Thus calls to `subclasses` get a
free performance boost when the class extends DescendantsTracker.

eugeneius commented:
Duck typing, aww yiss 🦆 ⌨️

Should we alias the other way too, so that `klass.subclasses` also gets faster when `DescendantsTracker` is mixed in?

Going further: maybe we could even rename `direct_descendants` to `subclasses` on `DescendantsTracker` and deprecate `direct_descendants`?

jonathanhefner commented:
> Should we alias the other way too, so that `klass.subclasses` also gets faster when `DescendantsTracker` is mixed in?

Sounds good! :rocket:

> Going further: maybe we could even rename `direct_descendants` to `subclasses` on `DescendantsTracker` and deprecate `direct_descendants`?

Agreed that `subclasses` is a nicer name, but don't we normally just hang on to aliases rather than deprecate them (since the maintenance cost is low)?

eugeneius commented:
I think it depends on the situation; in this case I'd like to avoid adding `Class#direct_descendants` since it will be available on every class by default. While `DescendantsTracker#direct_descendants` is public API, I doubt it's used much outside of Rails, so a deprecation should be relatively painless.

jonathanhefner commented:
@eugeneius I added a deprecation for `direct_descendants`.  I also renamed `@@direct_descendants` to `@@subclasses` for consistency, but I wanted you to :+1: the diff before I squashed commits.

eugeneius commented:
Looking at this again, I think you're right that we shouldn't deprecate `direct_descendants`: we can leave it as an alias on `DependencyTracker`, and just not add it to `Class`. That way `subclasses` is the 🦆, and `direct_descendants` is just there for backwards compatibility.

I think that means this can go back to being a one line implementation change (sorry! 😖).

jonathanhefner commented:
@eugeneius Lol, no problem.  Agreed!  :+1:

eugeneius commented:
Thank you! 🦆
